crates/rolldown/src/module_loader/normal_module_task.rs

```diff
@@ -132,35 +132,31 @@ impl NormalModuleTask {
     let resource_id = ResourceId::new(Arc::clone(&self.resolved_path.path));
     let stable_resource_id = resource_id.stabilize(&self.ctx.input_options.cwd);
 
-    self
-      .package_json
-      .as_ref()
-      .and_then(|p| {
-        p.check_side_effects_for(&stable_resource_id).map(DeterminedSideEffects::PackageJson)
-      })
-      .unwrap_or_else(|| {
-        let analyzed_side_effects = stmt_infos.iter().any(|stmt_info| stmt_info.side_effect);
-        DeterminedSideEffects::Analyzed(analyzed_side_effects)
-      });
-
-    let default_side_effects = self
-      .package_json
-      .as_ref()
-      .and_then(|p| {
-        p.check_side_effects_for(&stable_resource_id).map(DeterminedSideEffects::PackageJson)
-      })
-      .unwrap_or_else(|| {
-        let analyzed_side_effects = stmt_infos.iter().any(|stmt_info| stmt_info.side_effect);
-        DeterminedSideEffects::Analyzed(analyzed_side_effects)
-      });
+    // The side effects pripority is:
+    // 1. Hook side effects
+    // 2. Package.json side effects
+    // 3. Analyzed side effects
+    // We should skiping the `check_side_effects_for` if the hook side effects is not None.
+    let lazy_check_side_effects = || {
+      self
+        .package_json
+        .as_ref()
+        .and_then(|p| {
+          p.check_side_effects_for(&stable_resource_id).map(DeterminedSideEffects::UserDefined)
+        })
+        .unwrap_or_else(|| {
+          let analyzed_side_effects = stmt_infos.iter().any(|stmt_info| stmt_info.side_effect);
+          DeterminedSideEffects::Analyzed(analyzed_side_effects)
+        })
+    };
 
     let side_effects = match hook_side_effects {
       Some(side_effects) => match side_effects {
-        HookSideEffects::True => default_side_effects,
-        HookSideEffects::False => DeterminedSideEffects::PackageJson(false),
+        HookSideEffects::True => DeterminedSideEffects::UserDefined(true),
```

Comment:
Since side effect from `hook` has higher priority, we should apply the value directly.